### PR TITLE
[Enhancement] optimize analyze large column table take FE oom

### PR DESCRIPTION
**Pull request by Seaven**
## Why I'm doing:

when we create 35k columns table, and execute full analyze or sample analzye.

we will get a sql like:
```
insert statistics...
select column_1, row_count, min, max from table 
union all
select column_2, row_count, min, max from table 
union all
select column_3, row_count, min, max from table
....
```

the table process in Optimizer:
```
1. analyze table, storage all columns in QueryAnalzyer
2. transform table to scan operator, create columnRef from all columns
3. prune columns in Optimizer
```

the step 1&2 is necessary, because we don't know what columns will used in other-clause, but it's will cost large memory to copy columns

and in statistics SQL, it's a simple format sql, we can try to prune columns in analyze step

and other bottleneck point:
img_v3_02br_712f1eac-956c-4ad3-8e3d-2c4647893b5g
img_v3_02br_fabef1e3-0f5e-4636-a532-c0e725f8809g

for fast schema change feature, FE need send whole schema to BE, will cost large memory and time...

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -463,12 +463,12 @@ private OptExpression logicalRuleRewrite(
 
         ruleRewriteIterative(tree, rootTaskContext, new MergeTwoProjectRule());
         ruleRewriteOnlyOnce(tree, rootTaskContext, RuleSetType.ELIMINATE_GROUP_BY);
-        ruleRewriteOnlyOnce(tree, rootTaskContext, EliminateAggRule.getInstance());
         ruleRewriteOnlyOnce(tree, rootTaskContext, new PushDownPredicateRankingWindowRule());
 
         ruleRewriteOnlyOnce(tree, rootTaskContext, new ConvertToEqualForNullRule());
         ruleRewriteOnlyOnce(tree, rootTaskContext, RuleSetType.PRUNE_COLUMNS);
         ruleRewriteIterative(tree, rootTaskContext, RuleSetType.PRUNE_UKFK_JOIN);
+        ruleRewriteOnlyOnce(tree, rootTaskContext, EliminateAggRule.getInstance());
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:471]: why need move this?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java]: I want to move the execute place after `PRUNE_COLUMNS`, the rule will check all columns, I think it's unnecessary

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java [hunk 2]
```diff
@@ -53,6 +54,8 @@ public enum TableHint {
 
     private Map<Expr, SlotRef> generatedExprToColumnRef = new HashMap<>();
 
+    private List<String> guessScanColumns = Collections.emptyList();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java:57]: I think scan columns  should be determinate,  not guessed.  could we rename it to prunedScanColumns?  which is consistent with your session variable name:  enable_analyze_phase_prune_columns

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java [hunk 3]
```diff
@@ -212,6 +213,7 @@ public Scope visitSelect(SelectRelation selectRelation, Scope scope) {
                 throw unsupportedException("Table function must be used with lateral join");
             }
             selectRelation.setRelation(resolvedRelation);
+            guessScanColumns(selectRelation, resolvedRelation);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java:216]: would better comment why we need to prune scan columns here.

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java:57]: not all case can be pruned, we can't compute the complate relation now, just can be guess....

**Review by kangkaisen — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by Seaven** [comment]
@mergify backport branch-3.2 branch-3.3

**Pull request merged**